oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneDocumentMaker.java

```diff
@@ -315,6 +313,38 @@ protected boolean indexTypeOrderedFields(Document doc, String pname, int tag, Pr
         }
         return fieldAdded;
     }
+    
+    protected static BytesRef checkTruncateLength(String prop, String value, String path, int maxLength) {
```

Comment:
Could we change the name of the function to `truncateToMaxLength`, `getTruncatedBytesRef` something like that? IMO `checkTruncateLength` suggests that the function might just be verifying something rather than actually performing a truncation operation.

Reply:
`checkTruncateLength` is meant to be `checkAndTruncate` and I removd And for brevity. So, maybe `checkAndTruncateLength` if its more clearer

Reply:
I don't think that splitting will make it clearer though. The method has the check interwined also to check the higher code points are not split.

Reply:
I think the function does quite a lot of things - and I generally like our code to follow the SRP. Some minor refactoring can lead to this, which IMO is a bit clearer: 

```java
// TODO: explain why we use 10
private static final int EMERGENCY_TRUNCATION_LENGTH = 10;
...
// assuming we rename checkTruncateLength() -> truncateToMaxLength()
protected static BytesRef truncateToMaxLength(String prop, String value, String path, int maxLength) {
    log.trace("Property {} at path:[{}] has value {}", prop, path, value);

    BytesRef ref = new BytesRef(value);
    if (ref.length <= maxLength) {
        return ref;
    }

    ref = truncateBytesRef(ref, maxLength);
    
    if (ref.length > maxLength) {
        ref = emergencyTruncate(ref, maxLength);
    }

    return ref;
}

private static BytesRef truncateBytesRef(BytesRef ref, int maxLength) {
    log.info("Truncating as length after encoding {} is > {} ", ref.length, maxLength);
    
    int end = calculateEndPosition(ref, maxLength);
    byte[] truncatedBytes = Arrays.copyOf(ref.bytes, end + 1);
    
    BytesRef truncatedRef = new BytesRef(new String(truncatedBytes, StandardCharsets.UTF_8));
    log.trace("Truncated to {}", truncatedRef.utf8ToString());
    
    return truncatedRef;
}

// TODO: explain this more in detail, with examples and why it's necessary. 
private static int calculateEndPosition(BytesRef ref, int maxLength) {
    int end = maxLength - 1;

    while ((ref.bytes[end] & 0b11000000) == 0b10000000) {
        end--;
    }
    if ((ref.bytes[end] & 0b11000000) == 0b11000000) {
        end--;
    }

    return end;
}

// TODO: explain why this is necessary and with examples. 
private static BytesRef emergencyTruncate(BytesRef ref, int maxLength) {
    log.error("Truncation did not work: still {} bytes", ref.length);
    String truncated = ref.utf8ToString();

    while (ref.length > maxLength) {
        truncated = truncated.substring(0, truncated.length() - EMERGENCY_TRUNCATION_LENGTH);
        ref = new BytesRef(truncated);
    }

    return ref;
}
```

Reply:
Hmm...I think this takes the refactoring too far and in fact makes readability harder. Generally  if there was a need for reuse for some of the sub methods it would have made sense. Explanation of the code can happen through documentation as well.

Reply:
You know, I see where you're coming from about the code length. But I believe that while it might appear longer, it doesn't reduce readability. In fact, breaking the code down helps us describe and understand what each segment is doing, and more crucially, makes it easier to explain _why_ it's doing that.

Imagine we come across a pesky bug half a year from now and then stumble upon this piece of code:

```java
while ((ref.bytes[end] & 0b11000000) == 0b10000000) { 
    end--; 
} 

if ((ref.bytes[end] & 0b11000000) == 0b11000000) { 
    end--; 
}
```
Yeah I see that it skips over tails of utf-8 multi-byte sequences up to 3 bytes. But why? 

Thus, wouldn't it be so much easier if we immediately knew the reason behind this logic, instead of scratching our heads trying to decode it? That's why I believe refactoring it is so crucial - it makes it so much easier to explain _why_ we are doing it.